### ISSUE TITLE
with_cache not restoring cache very well

### ISSUE THREAD

**Issue by Peter-Darton-i2** [CircleCI-Public/maven-orb#44]
TL;DR: I'm getting a lot of cache misses with the `with_cache` command due to it not following CircleCI recommended practise w.r.t. caching.

**Orb Version**

1.4.1

**Describe the bug**

I'm working on a pom-only maven project and I'm getting no caching every time I (re)build.
[CircleCI recommended practise with caching](https://circleci.com/docs/caching/#restoring-cache) is to try a perfect cache restore and, if that's not possible, then to try a "partial cache restore" so that, even if the orb can't find an exact cache-hit, it can at least find something useful to reduce build times.
Unfortunately, this orb's `with_cache` command doesn't seem to follow CircleCI's official guidance about caching and this results in sub-optimal functionality

... but it's likely a very simple fix.

**To Reproduce**

1. Have a simple maven project that builds on CircleCI using [the `test` job](https://github.com/CircleCI-Public/maven-orb/blob/v1.4.1/src/jobs/test.yml).
1. Let CircleCI build it. Make a note of the output of the cache steps.
1. Make a simple change to the `pom.xml` file, even a non-functional change like editing a comment.
1. Let CircleCI rebuild it.
1. See that in the `restore_cache` step, no cache is restored.
   e.g. `No cache is found for key: maven-wuGcf15LD7WuwypI8EFl5DHA_rbQOnfx3NuH8iuE1H0=` where the key text doesn't match the key used in the previous build _and no attempt is made to fetch anything else_.

**Expected behavior**

In the situation where `pom.xml` has changed, the `restore_cache` step should fail to find an exact match (for e.g. `maven-wuGcf15LD7WuwypI8EFl5DHA_rbQOnfx3NuH8iuE1H0=`) but then continue on to do a "partial cache restore" using key `maven-`.
... because "something is better than nothing" and maven will discover any outdated/missing dependencies and update them.

(Ideally the `with_cache` command and every job that calls it should _also_ support a parameter that allows the user to invalidate their cache, e.g. have a `cache_version` parameter that defaults to `v1` and include that parameter value in the cache key too)

**Additional context**

In the CircleCI docs about caching, recommended practise is to specify multiple keys to "try" when restoring the cache, where each successive key is less specific than the previous.
The current code does not follow CircleCI recommended/documented practise and specifies a single key whose text is generated by checksuming the pom.xml file(s) ... which means that _any_ change to any pom.xml file will guarantee a total cache-miss on the next build.
What it should do instead is follow CircleCI recommended practise and specify multiple keys in [the restore_cache step](https://github.com/CircleCI-Public/maven-orb/blob/v1.4.1/src/commands/with_cache.yml#L42-L43).

i.e. instead of

```
  - restore_cache:
      key: maven-{{ checksum "/tmp/maven_cache_seed" }}
```

it should do

```
  - restore_cache:
      keys:
        - maven-{{ checksum "/tmp/maven_cache_seed" }}
        - maven-
```

(or, if we now have a new `cache_version` parameter defined, it should do
```
  - restore_cache:
      keys:
        - maven-<< parameters.cache_version >>-{{ checksum "/tmp/maven_cache_seed" }}
        - maven-<< parameters.cache_version >>-
```
)
